File: sigv4-signing/go/Sigv4Signer.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"github.com/gorilla/websocket"
	"io"
	"log"
	"net/http/httputil"
	"net/url"
	"os"
	"os/signal"
	"sort"
	"strings"
	"time"
)

const (
	ALGORITHM_AWS4_HMAC_SHA_256 = "AWS4-HMAC-SHA256"
	AWS4_REQUEST_TYPE           = "aws4_request"
	SERVICE                     = "kinesisvideo"
	X_AMZ_ALGORITHM             = "X-Amz-Algorithm"
	X_AMZ_CREDENTIAL            = "X-Amz-Credential"
	X_AMZ_DATE                  = "X-Amz-Date"
	X_AMZ_EXPIRES               = "X-Amz-Expires"
	X_AMZ_SECURITY_TOKEN        = "X-Amz-Security-Token"
	X_AMZ_SIGNATURE             = "X-Amz-Signature"
	X_AMZ_SIGNED_HEADERS        = "X-Amz-SignedHeaders"
	NEW_LINE_DELIMITER          = "\n"
	DATE_PATTERN                = "20060102"
	TIME_PATTERN                = "20060102T150405Z"
	METHOD                      = "GET"
	SIGNED_HEADERS              = "host"
)

type AwsV4Signer struct{}

// SigV4 Signer sample reference for the Kinesis Video Streams WebRTC WebSocket connections
// Uses Gorilla WebSockets Go Module

func main() {

	////////////////////////////////////////////
	// ***Sample connection test:*** - fill in the variables:

	// URI to sign.
	//
	// Connect as Master URL - GetSignalingChannelEndpoint (master role) + Query Parameters: Channel ARN as X-Amz-ChannelARN
	//   Ref: https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-websocket-apis-2.html
	// Connect as Viewer URL - GetSignalingChannelEndpoint (viewer role) + Query Parameters: Channel ARN as X-Amz-ChannelARN & Client Id as X-Amz-ClientId
	//   Ref: https://docs.aws.amazon.com/kinesisvideostreams-webrtc-dg/latest/devguide/kvswebrtc-websocket-apis-1.html
	//
	// Viewer URL example: wss://v-a1b2c3d4.kinesisvideo.us-west-2.amazonaws.com?X-Amz-ChannelARN=arn:aws:kinesisvideo:us-west-2:123456789012:channel/demo-channel/1234567890123&X-Amz-ClientId=d7d1c6e2-9cb0-4d61-bea9-ecb3d3816557
	//
	// **Note**: The Signaling Channel Endpoints are different, depending on the role (master/viewer) specified in GetSignalingChannelEndpoint API call.
	//   Ref: https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_SingleMasterChannelEndpointConfiguration.html#KinesisVideo-Type-SingleMasterChannelEndpointConfiguration-Role
	uri, _ := url.Parse("wss://<Your GetSignalingEndpoint response hostname>?X-Amz-ChannelARN=<YourChannelARN>&X-Amz-ClientId=<YourClientId>")

	// Secure WebSocket method "wss" plus hostname obtained from GetSignalingChannelEndpoint.
	// Viewer URL example: wss://v-a1b2c3d4.kinesisvideo.us-west-2.amazonaws.com
	//
	// **Note**: The Signaling Channel Endpoints are different, depending on the role (master/viewer) specified in GetSignalingChannelEndpoint API call.
	//   Ref: https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/API_SingleMasterChannelEndpointConfiguration.html#KinesisVideo-Type-SingleMasterChannelEndpointConfiguration-Role
	wssUri, _ := url.Parse("wss://<Your GetSignalingEndpoint Response hostname>")

	// Test Credentials for sample connection verification. SessionToken can be empty string if using non-temporary credentials.
	accessKey := "YourAccessKey"
	secretKey := "YourSecretKey"
	sessionToken := "YourSessionToken"

	// AWS Region (e.g. us-west-2)
	region := "YourChannelRegion"

	////////////////////////////////////////////


	// Compute SigV4 signing
	// Ref: https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
	// https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-android/blob/master/src/main/java/com/amazonaws/kinesisvideo/utils/AwsV4Signer.java

	signedUri := AwsV4Signer{}.sign(uri, accessKey, secretKey, sessionToken, wssUri, region)
	println("Go SignedURL log for verifications (e.g. using wscat ) =", signedUri.String())

	// Connect to the SigV4 signed Endpoint
	interrupt := make(chan os.Signal, 1)
	signal.Notify(interrupt, os.Interrupt)

	c, resp, err := websocket.DefaultDialer.Dial(signedUri.String(), nil)

	// Print the responses from the connection
	println("Response code=", resp.StatusCode)
	if err != nil {
		log.Fatalln("Error in connection: ", err)
	}

	b, errBodyResp := io.ReadAll(resp.Body)
	if errBodyResp != nil {
		log.Fatalln("Error in reading response body: ", errBodyResp)
	}

	fmt.Println("Response body", string(b))

	body, errResp := httputil.DumpResponse(resp, true)
	if errResp != nil {
		log.Fatal("Error printing response body:", err)
	}

	fmt.Println(string(body))

	defer c.Close()

}

func (signer AwsV4Signer) sign(uri *url.URL, accessKey, secretKey, sessionToken string, wssURI *url.URL, region string) *url.URL {

	dateMilli := time.Now().UnixMilli()
	amzDate := getTimeStamp(dateMilli)
	datestamp := getDateStamp(dateMilli)

	queryParamsMap := buildQueryParamsMap(uri, accessKey, sessionToken, region, amzDate, datestamp)
	canonicalQuerystring := getCanonicalizedQueryString(queryParamsMap)
	canonicalRequest := getCanonicalRequest(uri, canonicalQuerystring)
	stringToSign := signString(amzDate, createCredentialScope(region, datestamp), canonicalRequest)

	signatureKey := getSignatureKey(secretKey, datestamp, region, SERVICE)
	signature := hex.EncodeToString(hmacSha256(stringToSign, signatureKey))

	signedCanonicalQueryString := canonicalQuerystring + "&" + X_AMZ_SIGNATURE + "=" + signature

	signedURI, err := url.Parse(wssURI.Scheme + "://" + wssURI.Host + getCanonicalURI(uri) + "?" + signedCanonicalQueryString)
	if err != nil {
		fmt.Println(err)
	}

	return signedURI
}

func buildQueryParamsMap(uri *url.URL, accessKey, sessionToken, region, amzDate, datestamp string) map[string]string {

	xAmzCredential := urlEncode(accessKey + "/" + createCredentialScope(region, datestamp))

	// The SigV4 signer has a maximum time limit of five minutes.
	// Once a connection is established, peers exchange signaling messages,
	// and the P2P connection is successful, the media P2P session
	// can continue for longer period of time.
	queryParamsMap := map[string]string{
		X_AMZ_ALGORITHM:  ALGORITHM_AWS4_HMAC_SHA_256,
		X_AMZ_CREDENTIAL: xAmzCredential,
		X_AMZ_DATE:       amzDate,
		X_AMZ_EXPIRES:    "299",
	}

	if sessionToken != "" {
		queryParamsMap[X_AMZ_SECURITY_TOKEN] = urlEncode(sessionToken)
	}

	queryParamsMap[X_AMZ_SIGNED_HEADERS] = SIGNED_HEADERS

	if uri.RawQuery != "" {
		params := strings.Split(uri.RawQuery, "&")
		for _, param := range params {
			index := strings.Index(param, "=")
			if index > 0 {
				queryParamsMap[param[:index]] = urlEncode(param[index+1:])
			}
		}
	}

	println("\n")

	return queryParamsMap
}

func createCredentialScope(region, datestamp string) string {
	ret := strings.Join([]string{datestamp, region, SERVICE, AWS4_REQUEST_TYPE}, "/")
	return ret
}

func getCanonicalRequest(uri *url.URL, canonicalQuerystring string) string {
	// https://pkg.go.dev/crypto/sha256
	payloadHash := sha256.New().Sum([]byte(""))
	canonicalUri := getCanonicalURI(uri)
	canonicalHeaders := "host:" + uri.Host + NEW_LINE_DELIMITER
	canonicalRequest := strings.Join([]string{METHOD, canonicalUri, canonicalQuerystring, canonicalHeaders, SIGNED_HEADERS, hex.EncodeToString(payloadHash)}, NEW_LINE_DELIMITER)

	return canonicalRequest
}

func getCanonicalURI(uri *url.URL) string {
	if uri.Path == "" {
		return "/"
	}
	return uri.Path
}

func signString(amzDate, credentialScope, canonicalRequest string) string {
	// https://pkg.go.dev/crypto/sha256#Sum256
	// Sum256 returns the SHA256 checksum of the data.
	result := sha256.Sum256([]byte(canonicalRequest))
	// Slice to make it to variable size for encoding
	result2 := hex.EncodeToString(result[:])

	stringToSign := strings.Join([]string{ALGORITHM_AWS4_HMAC_SHA_256, amzDate, credentialScope, result2},
		NEW_LINE_DELIMITER)
	return stringToSign
}

func urlEncode(str string) string {
	ret := url.QueryEscape(str)
	return ret
}

func hmacSha256(data string, key []byte) []byte {
	mac := hmac.New(sha256.New, key)
	mac.Write([]byte(data))
	ret := mac.Sum(nil)
	return ret
}

func getSignatureKey(key, dateStamp, regionName, serviceName string) []byte {
	kSecret := []byte("AWS4" + key)
	kDate := hmacSha256(dateStamp, kSecret)
	kRegion := hmacSha256(regionName, kDate)
	kService := hmacSha256(serviceName, kRegion)
	ret := hmacSha256(AWS4_REQUEST_TYPE, kService)
	return ret
}

func getTimeStamp(dateMilli int64) string {
	ret := time.UnixMilli(dateMilli).UTC().Format(TIME_PATTERN)
	return ret
}

func getDateStamp(dateMilli int64) string {
	ret := time.UnixMilli(dateMilli).UTC().Format(DATE_PATTERN)
	return ret
}

func getCanonicalizedQueryString(queryParamsMap map[string]string) string {
	var queryKeys []string
	for key := range queryParamsMap {
		queryKeys = append(queryKeys, key)
	}
	sort.Strings(queryKeys)

	var builder strings.Builder
	for i, key := range queryKeys {
		builder.WriteString(key + "=" + queryParamsMap[key])
		if i < len(queryKeys)-1 {
			builder.WriteString("&")
		}
	}

	return builder.String()
}

File: sigv4-signing/rust/src/sigv4_signer.rs
/// Custom SigV4 Signer for KVS Signaling WebSocket, ConnectAsMaster and ConnectAsViewer
/// Compared to AWS SDK Rust implementation, this has less dependencies
pub mod aws_v4_signer {
    use crate::sigv4_signer_constants::constants::{
        ALGORITHM_AWS4_HMAC_SHA_256, AWS4_REQUEST_TYPE, DATE_PATTERN, METHOD, NEW_LINE_DELIMITER,
        SERVICE, SIGNED_HEADERS, TIME_PATTERN, X_AMZ_ALGORITHM, X_AMZ_CREDENTIAL, X_AMZ_DATE,
        X_AMZ_EXPIRES, X_AMZ_SECURITY_TOKEN, X_AMZ_SIGNATURE, X_AMZ_SIGNED_HEADERS,
    };
    use chrono::DateTime;
    use hmac::digest::Digest;
    use hmac::{Hmac, Mac};
    use sha2::Sha256;
    use std::collections::HashMap;
    use url::{form_urlencoded, Url};

    pub struct AwsV4Signer {}

    impl AwsV4Signer {
        pub fn sign(
            uri: &Url,
            access_key: &str,
            secret_key: &str,
            session_token: &str,
            region: &str,
            date_milli: i64,
        ) -> Url {
            // Step 1. Create canonical request.
            let amz_date = Self::get_time_stamp(date_milli);
            let datestamp = Self::get_date_stamp(date_milli);
            let query_params_map = Self::build_query_params_map(
                uri,
                access_key,
                session_token,
                region,
                &amz_date,
                &datestamp,
            );
            let canonical_querystring = Self::get_canonicalized_query_string(&query_params_map);
            let canonical_request = Self::get_canonical_request(uri, &canonical_querystring);

            // Step 2. Construct StringToSign.
            let string_to_sign = Self::sign_string(
                &amz_date,
                Self::create_credential_scope(region, datestamp.as_str()).as_str(),
                &canonical_request,
            );

            // Step 3. Calculate the signature.
            let signature_key = Self::get_signature_key(secret_key, &datestamp, region, SERVICE);
            let signature = hex::encode(Self::hmac_sha256(string_to_sign.as_str(), &signature_key));

            // Step 4. Combine steps 1 and 3 to form the final URL.
            let signed_canonical_query_string = format!(
                "{}&{}={}",
                canonical_querystring, X_AMZ_SIGNATURE, signature
            );
            let mut signed_uri = uri.clone();
            signed_uri.set_query(Some(&signed_canonical_query_string));

            signed_uri
        }
        pub fn build_query_params_map(
            uri: &Url,
            access_key: &str,
            session_token: &str,
            region: &str,
            amz_date: &str,
            datestamp: &str,
        ) -> HashMap<String, String> {
            let x_amz_credential = Self::url_encode(&format!(
                "{}/{}",
                access_key,
                Self::create_credential_scope(region, datestamp)
            ));

            let mut query_params_map = HashMap::new();
            query_params_map.insert(
                X_AMZ_ALGORITHM.to_string(),
                ALGORITHM_AWS4_HMAC_SHA_256.to_string(),
            );
            query_params_map.insert(X_AMZ_CREDENTIAL.to_string(), x_amz_credential);
            query_params_map.insert(X_AMZ_DATE.to_string(), amz_date.to_string());
            query_params_map.insert(X_AMZ_EXPIRES.to_string(), "299".to_string());

            if !session_token.is_empty() {
                query_params_map.insert(
                    X_AMZ_SECURITY_TOKEN.to_string(),
                    Self::url_encode(session_token),
                );
            }

            query_params_map.insert(X_AMZ_SIGNED_HEADERS.to_string(), SIGNED_HEADERS.to_string());

            // Add the query parameters included in the uri.
            // Note: query parameters follow the format: key1=val1&key2=val2&key3=val3
            if let Some(query) = uri.query() {
                for param in query.split('&') {
                    if let Some((key, value)) = param.split_once('=') {
                        query_params_map.insert(key.to_string(), Self::url_encode(value));
                    }
                }
            }

            query_params_map
        }

        pub fn create_credential_scope(region: &str, datestamp: &str) -> String {
            vec![datestamp, region, SERVICE, AWS4_REQUEST_TYPE].join("/")
        }

        pub fn get_canonical_request(uri: &Url, canonical_querystring: &str) -> String {
            // Compute SHA-256 hash of an empty payload
            let payload_hash = Sha256::digest(&[] as &[u8]);
            let payload_hash_hex = hex::encode(payload_hash);

            // Get canonical URI
            let canonical_uri = Self::get_canonical_uri(uri);

            // Build canonical headers
            let canonical_headers = format!(
                "host:{}{}",
                uri.host_str().unwrap_or(""),
                NEW_LINE_DELIMITER
            );

            vec![
                METHOD,
                &*canonical_uri,
                canonical_querystring,
                &*canonical_headers,
                SIGNED_HEADERS,
                &*payload_hash_hex,
            ]
            .join(NEW_LINE_DELIMITER)
        }

        pub fn get_canonical_uri(uri: &Url) -> String {
            if uri.path().is_empty() {
                "/".to_string()
            } else {
                uri.path().to_string()
            }
        }

        pub fn sign_string(
            amz_date: &str,
            credential_scope: &str,
            canonical_request: &str,
        ) -> String {
            let result_hex = hex::encode(Sha256::digest(canonical_request.as_bytes()));

            format!(
                "{}\n{}\n{}\n{}",
                ALGORITHM_AWS4_HMAC_SHA_256, amz_date, credential_scope, result_hex
            )
        }

        pub fn url_encode(s: &str) -> String {
            form_urlencoded::byte_serialize(s.as_bytes()).collect()
        }

        pub fn hmac_sha256(data: &str, key: &[u8]) -> Vec<u8> {
            let mut hmac: Hmac<Sha256> = Hmac::new_from_slice(key).unwrap();
            hmac.update(data.as_bytes());
            hmac.finalize().into_bytes().to_vec()
        }

        pub fn get_signature_key(
            key: &str,
            date_stamp: &str,
            region_name: &str,
            service_name: &str,
        ) -> Vec<u8> {
            let binding = format!("AWS4{}", &key).to_string();
            let k_secret = binding.as_bytes();
            let k_date = Self::hmac_sha256(&date_stamp, &k_secret);
            let k_region = Self::hmac_sha256(&region_name, &k_date);
            let k_service = Self::hmac_sha256(&service_name, &k_region);
            Self::hmac_sha256(&AWS4_REQUEST_TYPE, &k_service)
        }

        pub fn get_time_stamp(date_milli: i64) -> String {
            let datetime = DateTime::from_timestamp_millis(date_milli).unwrap();
            datetime.format(&TIME_PATTERN).to_string()
        }

        pub fn get_date_stamp(date_milli: i64) -> String {
            let datetime = DateTime::from_timestamp_millis(date_milli).unwrap();
            datetime.format(&DATE_PATTERN).to_string()
        }

        pub fn get_canonicalized_query_string(params: &HashMap<String, String>) -> String {
            let mut query_pairs: Vec<String> =
                params.iter().map(|(k, v)| format!("{}={}", k, v)).collect();
            query_pairs.sort(); // Must be in canonical order
            query_pairs.join("&")
        }
    }
}
